### PR TITLE
Refactor `TFSwinLayer` to increase serving compatibility

### PR DESCRIPTION
**Pull request by harrydrippin**
# What does this PR do?

This PR refactors two parts of `TFSwinLayer` like below, to increase serving compatibility on multiple accelerators.

### Fix incompatible type cast on several serving architecture

On the `window_reverse` function, there is incompatible computation on the graph when restoring batch size:

```python
x = shape_list(windows)[0]
y = tf.cast(height * width / (window_size * window_size), tf.int32)
batch_size = int(x / y) # <- Here
```

This operation can cause tracing failures depending on the accelerator or compiler SDKs. I have confirmed that AWS Neuron SDK (w/ Inferentia) cannot trace this part properly.

We have `batch_size` variable on graph inside `call()`, so I modified this function to use this value by passing it as a function argument.

### Fix mixed use of member & local variables

Several parts are using window size, but some use `window_size`, and some use `self.window_size`. This can cause tracing failure because of unclear separation between compile-time constants and runtime values.

This PR applies to fix this by not using `self.window_size` on everywhere, but only determining the actual window size on `call()`.

## Review

This PR is related with Swin Transformer and TensorFlow.
R: @amyeroberts

### CODE REVIEW

File: src/transformers/models/swin/modeling_tf_swin.py [hunk 1]
```diff
@@ -222,13 +222,10 @@ def window_partition(input_feature: tf.Tensor, window_size: int) -> tf.Tensor:
...
 
-def window_reverse(windows: tf.Tensor, window_size: int, height: int, width: int) -> tf.Tensor:
+def window_reverse(windows: tf.Tensor, batch_size: int, window_size: int, height: int, width: int) -> tf.Tensor:
     """
     Merges windows to produce higher resolution features.
     """
-    x = shape_list(windows)[0]
-    y = tf.cast(height * width / (window_size * window_size), tf.int32)
-    batch_size = int(x / y)
     windows = tf.reshape(
```
Comment (amyeroberts) [inline src/transformers/models/swin/modeling_tf_swin.py:229]: I'd prefer it if we could keep the same signature for the pytorch and tensorflow module level functions. It's good you dug into this as it highlights a bug in my original implementation (using `int` here vs. `floor` in the pytorch version). 

What do you think about the following as an alternative?: 
```
def window_reverse(windows: tf.Tensor, window_size: int, height: int, width: int) -> tf.Tensor:
    """
    Merges windows to produce higher resolution features.
    """
    x = tf.shape(windows)[0]
    y = tf.cast(height * width / (window_size * window_size), tf.int32)
    batch_size = tf.math.floordiv(x, y)
    windows = tf.reshape(
        windows, (batch_size, height // window_size, width // window_size, window_size, window_size, -1)
    )
    windows = tf.transpose(windows, (0, 1, 3, 2, 4, 5))
    windows = tf.reshape(windows, (batch_size, height, width, -1))
    return windows
```

Comment (harrydrippin) [reply src/transformers/models/swin/modeling_tf_swin.py:229]: Great! I'll apply this version to this PR.

**Review by amyeroberts — APPROVED** [review]
LGTM! Thanks for doing so much investigative work and for this fix 🔧

**Review by sgugger — APPROVED** [review]
Thanks for fixing!

**Pull request merged**